File: 10-逆向工程/01-二进制逆向/ELF符号哈希查找/python/elf_hash.py
"""ELF 符号哈希查找：DT_HASH 与 DT_GNU_HASH。

复刻对象（原文实读）：
  * glibc master `elf/simple-dl-new-hash.h`  -> __simple_dl_new_hash（GNU hash）
  * glibc master `elf/simple-dl-hash.h`      -> __simple_dl_elf_hash（SysV hash）
  * glibc master `elf/dl-setup_hash.c`       -> _dl_setup_hash（表头布局）
  * glibc master `elf/dl-lookup.c`           -> do_lookup_x（bloom 过滤 + chain 扫描）
  * glibc master `sysdeps/generic/ldsodefs.h`-> ELF_MACHINE_HASH_SYMIDX 默认定义

关键事实（全部来自上述源码，不是推测）：
  1. GNU hash: h = 5381; h = h*33 + c（uint32 回绕）
  2. SysV hash: h = (h<<4) + c; hi = h & 0xf0000000; h ^= hi>>24; h &= 0x0fffffff
  3. DT_GNU_HASH 表头 4 个 uint32：nbuckets / symbias / bitmask_nwords / shift；
     bitmask_nwords 必须是 2 的幂（源码里直接 assert）
  4. bloom 字下标 = (hash / __ELF_NATIVE_CLASS) & (nwords-1)，注意是**除法**不是移位
  5. 判据 (w >> (h&63)) & (w >> ((h>>shift)&63)) & 1
  6. chain 项：((hv ^ h) >> 1) == 0 才是候选（最低位被让给了 chain 终止位）
  7. symidx = hasharr - chain_zero（默认 ELF_MACHINE_HASH_SYMIDX）
"""

MASK32 = 0xFFFFFFFF
NATIVE_CLASS = 64          # __ELF_NATIVE_CLASS，ELF64 下为 64
CLASS_BITS = 6             # log2(64)，用于 hash & (__ELF_NATIVE_CLASS-1)
POW2_MINUS_1 = NATIVE_CLASS - 1

UNDEF = 0                  # STN_UNDEF


# ---------------------------------------------------------------- 哈希函数

def dl_elf_hash(name):
    """SysV DT_HASH 用：glibc __simple_dl_elf_hash。"""
    if isinstance(name, str):
        name = name.encode()
    h = 0
    for c in name:
        h = (h << 4) + c
        hi = h & 0xF0000000
        h ^= hi >> 24
        h &= 0x0FFFFFFF
    return h & MASK32


def dl_new_hash(name):
    """GNU DT_GNU_HASH 用：glibc __simple_dl_new_hash（djb2 变体）。"""
    if isinstance(name, str):
        name = name.encode()
    h = 5381
    for c in name:
        h = (h * 33 + c) & MASK32
    return h


# ---------------------------------------------------------------- 建表

class GnuHashTable:
    """按真实 linker 的口径构造一张 .gnu.hash。

    symbols: 完整符号表（下标 = 符号索引）
    symbias: 前 symbias 个符号不进哈希表（undefined / local），与 DT_GNU_HASH 的
             第二个字同义
    """

    def __init__(self, symbols, symbias=1, nbuckets=4, nwords=2, shift=6):
        if nwords & (nwords - 1) != 0:
            raise ValueError("bitmask_nwords must be a power of two")
        if nbuckets <= 0:
            raise ValueError("nbuckets must be positive")
        self.symbols = list(symbols)
        self.symbias = symbias
        self.nbuckets = nbuckets
        self.nwords = nwords
        self.idxbits = nwords - 1
        self.shift = shift

        exported = list(range(symbias, len(self.symbols)))
        # linker 会按 bucket 排序让同一 bucket 的符号在符号表里连成一段
        exported.sort(key=lambda i: dl_new_hash(self.symbols[i]) % nbuckets)
        order = {i: k for k, i in enumerate(exported)}
        # 重排符号表本体，使 chain 连续（真实 .so 就是这个形态）
        head = self.symbols[:symbias]
        self.symbols = head + [self.symbols[i] for i in exported]
        self.exported = list(range(symbias, len(self.symbols)))

        self.buckets = [0] * nbuckets
        self.chain_zero = [0] * len(self.symbols)
        self.bitmask = [0] * nwords

        for pos, i in enumerate(self.exported):
            h = dl_new_hash(self.symbols[i])
            b = h % nbuckets
            if self.buckets[b] == 0:
                self.buckets[b] = i
            nxt = self.exported[pos + 1] if pos + 1 < len(self.exported) else None
            last = nxt is None or dl_new_hash(self.symbols[nxt]) % nbuckets != b
            self.chain_zero[i] = (h & ~1) | (1 if last else 0)
            w = (h // NATIVE_CLASS) & self.idxbits
            self.bitmask[w] |= (1 << (h & POW2_MINUS_1))
            self.bitmask[w] |= (1 << ((h >> self.shift) & POW2_MINUS_1))

    # ------------------------------------------------------------ 原始字节

    def words32(self):
        """序列化成 DT_GNU_HASH 指向的 uint32 数组（_dl_setup_hash 的逆）。"""
        w = [self.nbuckets, self.symbias, self.nwords, self.shift]
        for v in self.bitmask:
            w.append(v & MASK32)
            w.append((v >> 32) & MASK32)
        w.extend(self.buckets)
        w.extend(self.chain_zero[self.symbias:])
        return w

    # ------------------------------------------------------------ 查找

    def bloom_probe(self, h):
        """只做 bloom 这一步，返回 (命中?, 用到的字下标, bit1, bit2)。"""
        widx = (h // NATIVE_CLASS) & self.idxbits
        word = self.bitmask[widx]
        b1 = h & POW2_MINUS_1
        b2 = (h >> self.shift) & POW2_MINUS_1
        hit = ((word >> b1) & (word >> b2) & 1) == 1
        return hit, widx, b1, b2

    def lookup(self, name, trace=False):
        """复刻 do_lookup_x 的 GNU hash 分支。返回符号下标或 None。"""
        h = dl_new_hash(name)
        steps = []
        hit, widx, b1, b2 = self.bloom_probe(h)
        steps.append(("bloom", widx, b1, b2, hit))
        if not hit:
            return (None, steps) if trace else None
        bucket = self.buckets[h % self.nbuckets]
        steps.append(("bucket", h % self.nbuckets, bucket))
        if bucket == 0:
            return (None, steps) if trace else None
        i = bucket
        while True:
            hv = self.chain_zero[i]
            if ((hv ^ h) >> 1) == 0:
                steps.append(("candidate", i, self.symbols[i]))
                if self.symbols[i] == name:
                    return (i, steps) if trace else i
            if hv & 1:
                break
            i += 1
        return (None, steps) if trace else None

    def symidx_from_hasharr(self, bucket, offset):
        """ELF_MACHINE_HASH_SYMIDX 的默认实现：hasharr - l_gnu_chain_zero。

        chain_zero 以符号索引为下标，所以「链上第 offset 项」的符号索引就是
        bucket + offset —— 这也是为什么 l_gnu_chain_zero 要减掉 symbias。
        """
        return bucket + offset


class SysvHashTable:
    """DT_HASH：nbuckets / nchain / buckets[] / chain[]。"""

    def __init__(self, symbols, symbias=1, nbuckets=4):
        self.symbols = list(symbols)
        self.symbias = symbias
        self.nbuckets = nbuckets
        nchain = len(self.symbols)
        self.buckets = [UNDEF] * nbuckets
        self.chain = [UNDEF] * nchain
        # chain 是「同 bucket 的下一个符号下标」，倒序插入形成链表
        for i in range(nchain - 1, symbias - 1, -1):
            b = dl_elf_hash(self.symbols[i]) % nbuckets
            self.chain[i] = self.buckets[b]
            self.buckets[b] = i

    def words32(self):
        w = [self.nbuckets, len(self.symbols)]
        w.extend(self.buckets)
        w.extend(self.chain)
        return w

    def lookup(self, name, trace=False):
        h = dl_elf_hash(name)
        steps = [("hash", h)]
        y = self.buckets[h % self.nbuckets]
        steps.append(("bucket", h % self.nbuckets, y))
        while y != UNDEF:
            steps.append(("candidate", y, self.symbols[y]))
            if self.symbols[y] == name:
                return (y, steps) if trace else y
            y = self.chain[y]
        return (None, steps) if trace else None


# ---------------------------------------------------------------- 解析

def parse_gnu_hash(words):
    """复刻 _dl_setup_hash 的 DT_GNU_HASH 分支。"""
    nbuckets = words[0]
    symbias = words[1]
    nwords = words[2]
    if nwords & (nwords - 1) != 0:
        raise ValueError("bitmask_nwords must be a power of two")
    idxbits = nwords - 1
    shift = words[3]
    off = 4
    bitmask = []
    for _ in range(nwords):
        lo = words[off]
        hi = words[off + 1]
        bitmask.append((hi << 32) | lo)
        off += 2
    buckets = words[off:off + nbuckets]
    off += nbuckets
    # l_gnu_chain_zero = hash32 - symbias：下标直接就是符号索引
    chain_zero = [0] * symbias + list(words[off:off + len(words) - off])
    return {
        "nbuckets": nbuckets,
        "symbias": symbias,
        "nwords": nwords,
        "idxbits": idxbits,
        "shift": shift,
        "bitmask": bitmask,
        "buckets": buckets,
        "chain_zero": chain_zero,
    }


def parse_sysv_hash(words):
    nbuckets = words[0]
    nchain = words[1]
    return {
        "nbuckets": nbuckets,
        "nchain": nchain,
        "buckets": words[2:2 + nbuckets],
        "chain": words[2 + nbuckets:2 + nbuckets + nchain],
    }

File: 10-逆向工程/01-二进制逆向/ELF符号哈希查找/go/elf_hash.go
// Package elfhash 复刻 glibc 的 ELF 符号哈希查找：DT_HASH 与 DT_GNU_HASH。
//
// 原文实读：
//
//	elf/simple-dl-new-hash.h   GNU hash: h=5381; h=h*33+c
//	elf/simple-dl-hash.h       SysV hash: h=(h<<4)+c; hi=h&0xf0000000; ...
//	elf/dl-setup_hash.c        表头 4 字 + bitmask_nwords 必须是 2 的幂
//	elf/dl-lookup.c            bloom 过滤 + chain 扫描
//	sysdeps/generic/ldsodefs.h ELF_MACHINE_HASH_SYMIDX = hasharr - chain_zero
package main

import "errors"

const nativeClass = 64 // __ELF_NATIVE_CLASS (ELF64)
const classMask = nativeClass - 1

// DlElfHash 是 SysV DT_HASH 用的哈希，结果恒小于 2^28。
func DlElfHash(name string) uint32 {
	var h uint32
	for i := 0; i < len(name); i++ {
		h = (h << 4) + uint32(name[i])
		hi := h & 0xf0000000
		h ^= hi >> 24
		h &= 0x0fffffff
	}
	return h
}

// DlNewHash 是 GNU DT_GNU_HASH 用的哈希（djb2 变体，uint32 回绕）。
func DlNewHash(name string) uint32 {
	h := uint32(5381)
	for i := 0; i < len(name); i++ {
		h = h*33 + uint32(name[i])
	}
	return h
}

// GnuHashTable 按 linker 口径构造一张 .gnu.hash。
type GnuHashTable struct {
	Symbols   []string
	Symbias   int
	Nbuckets  int
	Nwords    int
	Idxbits   int
	Shift     uint
	Bitmask   []uint64
	Buckets   []uint32
	ChainZero []uint32
	Exported  []int
}

// NewGnuHashTable 依据 symbols 建表；symbias 之前的符号不进哈希表。
func NewGnuHashTable(symbols []string, symbias, nbuckets, nwords int, shift uint) (*GnuHashTable, error) {
	if nwords&(nwords-1) != 0 {
		return nil, errors.New("bitmask_nwords must be a power of two")
	}
	if nbuckets <= 0 || symbias < 0 || symbias > len(symbols) {
		return nil, errors.New("bad table geometry")
	}

	exp := make([]int, 0, len(symbols))
	for i := symbias; i < len(symbols); i++ {
		exp = append(exp, i)
	}
	// 按 bucket 排序，使同一 bucket 的符号在符号表里连成一段
	sortByBucket(exp, symbols, nbuckets)

	head := make([]string, symbias)
	copy(head, symbols[:symbias])
	syms := make([]string, 0, len(symbols))
	syms = append(syms, head...)
	for _, i := range exp {
		syms = append(syms, symbols[i])
	}
	exported := make([]int, 0, len(exp))
	for i := symbias; i < len(syms); i++ {
		exported = append(exported, i)
	}

	t := &GnuHashTable{
		Symbols:   syms,
		Symbias:   symbias,
		Nbuckets:  nbuckets,
		Nwords:    nwords,
		Idxbits:   nwords - 1,
		Shift:     shift,
		Bitmask:   make([]uint64, nwords),
		Buckets:   make([]uint32, nbuckets),
		ChainZero: make([]uint32, len(syms)),
		Exported:  exported,
	}

	for pos, i := range exported {
		h := DlNewHash(syms[i])
		b := h % uint32(nbuckets)
		if t.Buckets[b] == 0 {
			t.Buckets[b] = uint32(i)
		}
		last := true
		if pos+1 < len(exported) {
			nxt := exported[pos+1]
			if DlNewHash(syms[nxt])%uint32(nbuckets) == b {
				last = false
			}
		}
		t.ChainZero[i] = (h & ^uint32(1)) | boolToU32(last)
		w := (h / nativeClass) & uint32(t.Idxbits)
		t.Bitmask[w] |= 1 << (h & classMask)
		t.Bitmask[w] |= 1 << ((h >> shift) & classMask)
	}
	return t, nil
}

func boolToU32(b bool) uint32 {
	if b {
		return 1
	}
	return 0
}

func sortByBucket(idx []int, symbols []string, nbuckets int) {
	// 插入排序：链短且顺序稳定，便于与 Python 侧逐项对照
	for a := 1; a < len(idx); a++ {
		v := idx[a]
		vb := DlNewHash(symbols[v]) % uint32(nbuckets)
		b := a - 1
		for b >= 0 && DlNewHash(symbols[idx[b]])%uint32(nbuckets) > vb {
			idx[b+1] = idx[b]
			b--
		}
		idx[b+1] = v
	}
}

// BloomProbe 只做 bloom 一步，返回是否放行以及用到的字下标与两个比特位。
func (t *GnuHashTable) BloomProbe(h uint32) (bool, uint32, uint, uint) {
	widx := (h / nativeClass) & uint32(t.Idxbits)
	word := t.Bitmask[widx]
	b1 := uint(h & classMask)
	b2 := uint((h >> t.Shift) & classMask)
	hit := ((word>>b1)&(word>>b2))&1 == 1
	return hit, widx, b1, b2
}

// Lookup 复刻 do_lookup_x 的 GNU hash 分支，返回符号下标（未找到返回 -1）。
func (t *GnuHashTable) Lookup(name string) int {
	h := DlNewHash(name)
	if hit, _, _, _ := t.BloomProbe(h); !hit {
		return -1
	}
	bucket := t.Buckets[h%uint32(t.Nbuckets)]
	if bucket == 0 {
		return -1
	}
	i := int(bucket)
	for {
		if i < 0 || i >= len(t.ChainZero) {
			return -1
		}
		hv := t.ChainZero[i]
		if ((hv ^ h) >> 1) == 0 && t.Symbols[i] == name {
			return i
		}
		if hv&1 == 1 {
			return -1
		}
		i++
	}
}

// SymidxFromHasharr 是 ELF_MACHINE_HASH_SYMIDX 的默认实现语义。
func (t *GnuHashTable) SymidxFromHasharr(bucket, offset int) int {
	return bucket + offset
}

// Words32 序列化为 DT_GNU_HASH 指向的 uint32 数组。
func (t *GnuHashTable) Words32() []uint32 {
	w := []uint32{uint32(t.Nbuckets), uint32(t.Symbias), uint32(t.Nwords), uint32(t.Shift)}
	for _, v := range t.Bitmask {
		w = append(w, uint32(v), uint32(v>>32))
	}
	w = append(w, t.Buckets...)
	for i := t.Symbias; i < len(t.ChainZero); i++ {
		w = append(w, t.ChainZero[i])
	}
	return w
}

// ParsedGnuHash 是 _dl_setup_hash 解析出的各字段。
type ParsedGnuHash struct {
	Nbuckets  uint32
	Symbias   uint32
	Nwords    uint32
	Idxbits   uint32
	Shift     uint32
	Bitmask   []uint64
	Buckets   []uint32
	ChainZero []uint32
}

// ParseGnuHash 复刻 _dl_setup_hash 的 DT_GNU_HASH 分支。
func ParseGnuHash(w []uint32) (*ParsedGnuHash, error) {
	if len(w) < 4 {
		return nil, errors.New("truncated gnu hash header")
	}
	nwords := w[2]
	if nwords&(nwords-1) != 0 {
		return nil, errors.New("bitmask_nwords must be a power of two")
	}
	p := &ParsedGnuHash{
		Nbuckets: w[0], Symbias: w[1], Nwords: nwords,
		Idxbits: nwords - 1, Shift: w[3],
	}
	off := 4
	for i := uint32(0); i < nwords; i++ {
		if off+1 >= len(w) {
			return nil, errors.New("truncated bitmask")
		}
		p.Bitmask = append(p.Bitmask, uint64(w[off])|uint64(w[off+1])<<32)
		off += 2
	}
	if off+int(p.Nbuckets) > len(w) {
		return nil, errors.New("truncated buckets")
	}
	p.Buckets = append(p.Buckets, w[off:off+int(p.Nbuckets)]...)
	off += int(p.Nbuckets)
	p.ChainZero = make([]uint32, int(p.Symbias))
	p.ChainZero = append(p.ChainZero, w[off:]...)
	return p, nil
}

// SysvHashTable 是 DT_HASH：nbuckets / nchain / buckets[] / chain[]。
type SysvHashTable struct {
	Symbols  []string
	Symbias  int
	Nbuckets int
	Buckets  []uint32
	Chain    []uint32
}

// NewSysvHashTable 建表；STN_UNDEF(0) 同时是空桶与链表尾的哨兵。
func NewSysvHashTable(symbols []string, symbias, nbuckets int) *SysvHashTable {
	t := &SysvHashTable{
		Symbols: symbols, Symbias: symbias, Nbuckets: nbuckets,
		Buckets: make([]uint32, nbuckets),
		Chain:   make([]uint32, len(symbols)),
	}
	for i := len(symbols) - 1; i >= symbias; i-- {
		b := DlElfHash(symbols[i]) % uint32(nbuckets)
		t.Chain[i] = t.Buckets[b]
		t.Buckets[b] = uint32(i)
	}
	return t
}

// Lookup 走 SysV 的 bucket→chain 链表。
func (t *SysvHashTable) Lookup(name string) int {
	h := DlElfHash(name)
	y := t.Buckets[h%uint32(t.Nbuckets)]
	for y != 0 {
		if t.Symbols[y] == name {
			return int(y)
		}
		y = t.Chain[y]
	}
	return -1
}

// Words32 序列化为 DT_HASH 指向的 uint32 数组。
func (t *SysvHashTable) Words32() []uint32 {
	w := []uint32{uint32(t.Nbuckets), uint32(len(t.Symbols))}
	w = append(w, t.Buckets...)
	w = append(w, t.Chain...)
	return w
}
